design/pool-configuration.md

```diff
@@ -0,0 +1,552 @@
+
+#### AddressPool
+
+The `AddressPool` is the entity responsible for the IP allocation part, regardless of how (and if) the ip is going to be advertised.
+
+Hence, we keep the same structure but remove the bgpAdvertisement section and the Protocol field.
+
+```go
+type addressPool struct {
+  Name              string
+  Addresses         []string
+  AvoidBuggyIPs     bool
+  AutoAssign        *bool
+}
+```
+
+Example CRD
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: AddressPool
+metadata:
+  name: addresspool-sample3
+  namespace: metallb-system
+spec:
+  addresses:
+    - 2002:2:2::1-2002:2:2::100
+    - 172.20.0.100/24
+  autoAssign: false
+  avoidBuggyIPs: true
+
+```
+
+#### BGP Advertisement
+
+The `BGPAdvertisement` entity represents the intention to advertise a given IP via the BGP Protocol. It contains properties related to the type of advertisement and a reference to the address pools we want we advertise. This initial version matches the current behaviour, where an IP is advertised to all the `BGPPeer`s configured in MetalLB.
+
+```go
+type bgpAdvertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+  AggregationLength *int
+  LocalPref         *uint32
+  Communities       []string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    peers:
+      - peer1
+      - peer2
+    - communities:
+        - 65535:65282
+    aggregationLength: 32
+    localPref: 100
+```
+
+A `BGPAdvertisement` with no pool selector is intended to select all the available `AddressPool`s.
+
+#### L2 Advertisement
+
+The `L2Advertisement` entity represents the intention to advertise a given IP via L2. Currently, there are no knobs related to L2 advertisement.
+
+```go
+type l2Advertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: lsadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+```
+
+A `L2Advertisement` with no pool selector is intended to select all the available `AddressPool`s.
+
+### BGP Peer
+
+The `BgpPeer` entity represents an external BGP peer we want to establish a session with. It contains the properties of the sessions, and it's a 1:1 mapping of the current bgp peer structure without the node selector. We also add the `bfdProfile` structure here, as it is part of the session definition.
+
+```go
+type bgpPeer struct {
+  Name              string
+  MyASN         uint32
+  ASN           uint32
+  Addr          string
+  SrcAddr       string
+  Port          uint16
+  HoldTime      string
+  KeepaliveTime string
+  RouterID      string
+  Password      string
+  BFDProfile    string
```

Comment:
I'm wondering if this should be a string, given that we're referencing a proper k8s API object here whose existence should likely be validated by a webhook upon creation of the BGP peer.

Reply:
Never mind, please disregard. I wasn't 100% clear on something here but now realize that all we need is reference a BFD profile name.